src/gui/processing/models/qgsmodelviewtoollink.cpp

```
@@ -0,0 +1,277 @@
+
+  // we need to manually pass this event down to items we want it to go to -- QGraphicsScene doesn't propagate
+  QList<QGraphicsItem *> items = scene()->items( event->modelPoint() );
+
+  QgsModelDesignerSocketGraphicItem *socket = nullptr;
+  for ( QGraphicsItem *item : items )
+  {
+    if ( ( socket = dynamic_cast<QgsModelDesignerSocketGraphicItem *>( item ) ) )
+    {
+      // snap
+      if ( mFromSocket != socket && mFromSocket->edge() != socket->edge() )
+      {
+        socket->modelHoverEnterEvent( event );
+        QPointF rubberEndPos = socket->mapToScene( socket->getPosition() );
+        mBezierRubberBand->update( rubberEndPos, Qt::KeyboardModifiers() );
+
+        break;
+      }
+    }
+  }
+
+  if ( socket == nullptr && mLastHoveredSocket != nullptr && socket != mLastHoveredSocket )
+  {
+    mLastHoveredSocket->modelHoverLeaveEvent( event );
+    mLastHoveredSocket = nullptr;
+  }
+  else
+  {
+    mLastHoveredSocket = socket;
+  }
+}
+
+void QgsModelViewToolLink::modelReleaseEvent( QgsModelViewMouseEvent *event )
+{
+  if ( event->button() != Qt::LeftButton )
+  {
+    return;
+  }
+  view()->setTool( mPreviousViewTool );
+  mBezierRubberBand->finish( event->modelPoint() );
+
+  // we need to manually pass this event down to items we want it to go to -- QGraphicsScene doesn't propagate
+  QList<QGraphicsItem *> items = scene()->items( event->modelPoint() );
+
+  mToSocket = nullptr;
+
+  for ( QGraphicsItem *item : items )
+  {
+    if ( QgsModelDesignerSocketGraphicItem *socket = dynamic_cast<QgsModelDesignerSocketGraphicItem *>( item ) )
+    {
+      mToSocket = socket;
+    }
+  }
+
+  // Do nothing if cursor don't land on another socket
+  if ( mToSocket == nullptr )
+  {
+    return;
+  }
+
+  // Do nothing if from socket and to socket are both input or both output
+  if ( mFromSocket->edge() == mToSocket->edge() )
+  {
+    return;
+  }
+
+  emit view() -> beginCommand( "Edit link" );
```

This is ugly -- externally emitting signals is not recommended. Better to add a public method to QgsModelGraphicsView and let that raise the command internally. 

(These public methods should ideally be named beginCommand/endCommand, with the existing signals being renamed to commandBegun/commandEnded)

---

Moved internally
 * renamed signals
 * added public method